cpp/src/arrow/compute/kernels/nth_to_indices.h

```diff
@@ -0,0 +1,53 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+class FunctionContext;
+
+/// \brief Returns indices that partition an array around n-th
+/// sorted element.
+///
+/// Find index of n-th(0 based) smallest value and perform indirect
+/// partition of an array around that element. Output indices[0 ~ n-1]
+/// holds values no greater than n-th element, and indices[n+1 ~ end]
+/// holds values no less than n-th element. Elements in each partition
+/// is not sorted. Nulls will be partitioned to the end of the output.
+/// Output is not guaranteed to be stable.
+///
+/// \param[in] ctx the FunctionContext
+/// \param[in] values array to sort
+/// \param[in] n n-th element to partition the array
+/// \param[out] offsets indices that would sort an array
+ARROW_EXPORT
+Status NthToIndices(FunctionContext* ctx, const Array& values, uint64_t n,
```

Comment:
I know `n` shouldn't be negative, but our convention is to use `int64_t` for lengths and offsets, so we should probably use that here too.

Reply:
Something else: you can return `Result<std::shared_ptr<Array>>` instead of Status + out-pointer.

Reply:
Changed n to int64.
I'm returning Status just to be consistent with other kernels. Result looks more convenient.